pkg/capabilities/consensus/ocr3/datafeeds/llo_aggregator.go

```diff
@@ -0,0 +1,409 @@
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize llo state: %w", err)
+	}
+
+	allStreamIDs := []uint32{}
+	for streamID := range currentState.StreamInfo {
+		allStreamIDs = append(allStreamIDs, streamID)
+	}
+	slices.Sort(allStreamIDs)
+	lggr.Debugw("determined streams to aggregate", "nStreamIds", len(allStreamIDs))
+
+	observationTimestamp, prices, err := lloStreamPrices(lggr, allStreamIDs, lloEvents, f)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get latest prices: %w", err)
+	}
+	lggr = logger.With(lggr, "observationTimestamp", observationTimestamp)
+
+	mustUpdateIDs := []uint32{}  // streamIDs that need to be updated per deviation or heartbeat
+	maybeUpdateIDs := []uint32{} // streamIDs that are within AllowedPartialStaleness percentage of their heartbeat
+	for _, streamID := range allStreamIDs {
+		previousStreamInfo := currentState.StreamInfo[streamID]
+		config := a.config.streams[streamID]
+		oldPrice := new(decimal.Decimal)
+		if uerr := oldPrice.UnmarshalBinary(previousStreamInfo.Price); uerr != nil {
+			lggr.Errorw("failed to unmarshal previous price", "streamID", streamID, "err", uerr)
+			continue
+		}
+		//oldPrice := big.NewInt(0).SetBytes(previousStreamInfo.Price)
+		newPrice := prices[streamID] //.Mul(decimal.NewFromInt(multiplier)).BigInt()
+		priceDeviation := deviationDecimal(*oldPrice, newPrice)
+		timeDiffNs := observationTimestamp - uint64(previousStreamInfo.Timestamp) //nolint:gosec // G115
+		lggr.Debugw("checking deviation and heartbeat",
+			"streamID", streamID,
+			"observationNs", observationTimestamp,
+			"perviousNs", previousStreamInfo.Timestamp,
+			"currStalenessNs", timeDiffNs,
+			"heartbeatSec", config.Heartbeat,
+			"oldPrice", oldPrice,
+			"newPrice", newPrice,
+			"currDeviation", priceDeviation,
+			"deviation", config.Deviation.InexactFloat64(),
+		)
+		if timeDiffNs > uint64(config.HeartbeatNanos()) || //nolint:gosec // G115
+			priceDeviation > config.Deviation.InexactFloat64() {
+			// this stream needs an update
+			previousStreamInfo.Timestamp = int64(observationTimestamp) //nolint: gosec // G115
+			var err2 error
+			previousStreamInfo.Price, err2 = prices[streamID].MarshalBinary()
+			if err2 != nil {
+				lggr.Errorw("failed to marshal price", "streamID", streamID, "err", err2)
+				continue
+			}
+			mustUpdateIDs = append(mustUpdateIDs, streamID)
+		} else if float64(timeDiffNs) > float64(config.HeartbeatNanos())*(1.0-a.config.allowedPartialStaleness) {
+			maybeUpdateIDs = append(maybeUpdateIDs, streamID)
+		}
+	}
+
+	// optimization that allows for more efficient batching
+	// if there is at least one stream that actually hit its deviation or heartbeat threshold,
+	// append all others that were within AllowedPartialStaleness percentage of their heartbeat
+	if len(mustUpdateIDs) > 0 {
+		mustUpdateIDs = append(mustUpdateIDs, maybeUpdateIDs...)
+		// deterministic order
+		slices.Sort(mustUpdateIDs)
```

Comment:
Technically not needed - the order will still be the same across nodes as long as allStreamIDs was sorted.